src/jv_type_number.c

```diff
@@ -0,0 +1,309 @@
+#include <stdint.h>
+#include <stddef.h>
+#include <assert.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdarg.h>
+#include <limits.h>
+#include <math.h>
+#include <float.h>
+
+#include "jv_type_private.h"
+#include "jv_dtoa.h"
+
+// this means that we will manage the space for the struct
+#define DECNUMDIGITS 1
+
+#include "decNumber/decNumber.h"
+
+/*
+ * Numbers
+ */
+
+enum {
+  JVP_NUMBER_NATIVE = 0,
+  JVP_NUMBER_DECIMAL = 1
+};
+
+#define JV_NUMBER_SIZE_INIT      (0)
+#define JV_NUMBER_SIZE_CONVERTED (1)
+
+#define JVP_FLAGS_NUMBER_NATIVE       JVP_MAKE_FLAGS(JV_KIND_NUMBER, JVP_MAKE_PFLAGS(JVP_NUMBER_NATIVE, 0))
+#define JVP_FLAGS_NUMBER_NATIVE_STR   JVP_MAKE_FLAGS(JV_KIND_NUMBER, JVP_MAKE_PFLAGS(JVP_NUMBER_NATIVE, 1))
+#define JVP_FLAGS_NUMBER_LITERAL      JVP_MAKE_FLAGS(JV_KIND_NUMBER, JVP_MAKE_PFLAGS(JVP_NUMBER_DECIMAL, 1))
+
+#define STR(x) #x
+#define XSTR(x) STR(x)
+#define DBL_MAX_STR XSTR(DBL_MAX)
+#define DBL_MIN_STR "-" XSTR(DBL_MAX)
+
+// the decimal precision of binary double
+#define BIN64_DEC_PRECISION  (17)
+#define DEC_NUMBER_STRING_GUARD (14)
+
+static decContext dec_context = {
+  DEC_MAX_DIGITS,
+  DEC_MAX_EMAX,
+  DEC_MIN_EMAX,
+  DEC_ROUND_HALF_UP,
+  0, /*no errors*/
+  0, /*status*/
+  0, /*no clamping*/
+};
+
+typedef struct {
+  jv_refcnt refcnt;
+  double num_double;
+  char * literal_data;
+  decNumber num_decimal; // must be the last field in the structure for memory management
+} jvp_literal_number;
+
+typedef struct {
+  decNumber number;
+  decNumberUnit units[1];
+} decNumberSingle;
+
+typedef struct {
+  decNumber number;
+  decNumberUnit units[BIN64_DEC_PRECISION];
+} decNumberDoublePrecision;
+
+static decContext dec_context_to_double = {
+  BIN64_DEC_PRECISION,
+  DEC_MAX_EMAX,
+  DEC_MIN_EMAX,
+  DEC_ROUND_HALF_UP,
+  0, /*no errors*/
+  0, /*status*/
+  0, /*no clamping*/
+};
+
+static struct {
+  uint64_t refcnt;
+  struct dtoa_context C;
+} shared_dtoa_context;
```

Comment:
I have concerns about these shared contexts and thread-safety. While the jq vm itself is single-threaded (and the main `jq` executable as well), we do expose a C API that other, multi-threaded programs could use. Two programs attempting to use the same contexts here could encounter issues.

I understand the purpose was to reduce allocations/improve performance, but we have to be careful with globals and shared state.

Reply:
OK. Indeed, this shared data will not be usable even with external syncrhonization techniques. 
I can change it to thread local storage. The only constraint here would be that the jv vm would need to be started and used and destroyed on the same thread.

Or, we could think of some "vm context" to be used anywhere inside the implementation its needed, but it looks like a much more impacting refactor, it would require passing the context everywhere around. But it would mean that a given vm instance could be used across multiple threads given some external sync to prevent simultaneous execution.

Reply:
I'd like to avoid the thread local storage so we don't have to pin a vm to its original thread. That seems like a pretty heavy cost on our users.

The vm context probably costs a bit more to solve than we'd like (certainly in implementation complexity).

That said, why do these need to be shared at all? Why not just make a `dtoa_context` or a `dec_context` on the stack when we need one? These don't seem to be particularly large (though the `dtoa_context` could be sneaky if it has to allocate a lot during processing). I'd like to investigate the performance cost of just making these on the stack every time.

Reply:
Actually, I have just realised that I can come up with a thread local implementation immune to thread move.

The shared objects will be managed per thread, with refcount; once a new jv_number requires a shared context it received an initialized pointer. Then, every time it needs the contex it should pass the pointer and the system will compare it to the current thread's shared pointer. If it's different then it will reduce the refcount for the other thread's context and increment the refcount for this thread's context, and return this thread's context, so that the jv_number already has a different pointer to work with currently.
this way, the refcount will properly flow between threads and the contexts will be freed once the last object is deallocated  or moved from that thread